### PR TITLE
feat(api): add is topic notification feature flag

### PR DESCRIPTION
**Pull request by p-fernandez**
### What change does this PR introduce?

Moves the feature flag that enables the topic notifications to work to Launch Darkly with its fallbacks to the environment variable flag and the hadcode.

### Why was this change needed?

This task is an example on how to migrate feature flags to Launch Darkly, keeping the current functionality and without having to touch the existing implementation while keeping the tests working as expected.

### CODE REVIEW

File: apps/api/src/app/feature-flags/use-cases/get-feature-flag/get-feature-flag.use-case.ts [hunk 1]
```diff
@@ -33,6 +33,17 @@ export class GetFeatureFlag {
...
+  async isTopicNotificationEnabled(featureFlagCommand: FeatureFlagCommand): Promise<boolean> {
+    const value = process.env.FF_IS_TOPIC_NOTIFICATION_ENABLED;
+    const fallbackValue = true; // It is a permanent feature now
+    const defaultValue = this.prepareBooleanStringFeatureFlag(value, fallbackValue);
+    const key = FeatureFlagsKeysEnum.IS_TOPIC_NOTIFICATION_ENABLED;
+
+    const command = this.buildCommand(featureFlagCommand, key, defaultValue);
+
+    return await this.execute(command);
+  }
```
Comment (p-fernandez) [inline apps/api/src/app/feature-flags/use-cases/get-feature-flag/get-feature-flag.use-case.ts:45]: Every feature flag can have its dedicated method to check the value or go to the fallback.

File: apps/api/src/app/feature-flags/e2e/get-feature-flag.e2e.ts [hunk 2]
```diff
@@ -72,13 +88,26 @@ describe('Get Feature Flag', () => {
...
-        expect(result).to.equal(true);
+      describe('IS_TOPIC_NOTIFICATION_ENABLED', () => {
+        it(`should get the feature flag value stored in Launch Darkly (enabled)
+           when the SDK key env variable is set regardless of the feature flag set`, async () => {
+          process.env.FF_IS_TOPIC_NOTIFICATION_ENABLED = 'false';
+
+          const result = await getFeatureFlag.isTemplateStoreEnabled(featureFlagCommand);
+
+          expect(result).to.equal(true);
+        });
```
Comment (p-fernandez) [inline apps/api/src/app/feature-flags/e2e/get-feature-flag.e2e.ts:110]: Testing the E2E to Launch Darkly. We should always test with the environment variable flag set to the opposite value of what's set in Launch Darkly in Test environment.

File: apps/api/src/app/events/usecases/map-trigger-recipients/map-trigger-recipients.use-case.ts [hunk 3]
```diff
@@ -86,11 +91,13 @@ export class MapTriggerRecipients {
     userId: UserId,
     recipients: TriggerRecipients
   ): Promise<ISubscribersDefine[]> {
-    /*
-     * TODO: We should manage the env variables from the config and not process.env
-     * https://github.com/motdotla/dotenv/issues/51
-     */
-    if (process.env.FF_IS_TOPIC_NOTIFICATION_ENABLED === 'true') {
```
Comment (p-fernandez) [inline apps/api/src/app/events/usecases/map-trigger-recipients/map-trigger-recipients.use-case.ts:93]: No need for spare checks in the environment process. We now can centralise everything in the feature flags module.

**Comment by scopsy** [comment]
Do we still need this feature flag in the system? It's been quite long before we released this functionality. If we still want to keep it, just need to validate the default behaviour for this it's Enabled, for cloud and self hosted

File: apps/api/src/app/events/usecases/map-trigger-recipients/map-trigger-recipients.e2e.ts [hunk 4]
```diff
@@ -605,8 +614,8 @@ const addSubscribersToTopic = async (
 
   const result = await topicRepository.findTopic(topicKey, _environmentId);
 
-  expect(result.subscribers.length).to.be.eql(subscribers.length);
-  expect(result.subscribers).to.have.members(subscribers.map((subscriber) => subscriber.subscriberId));
+  expect(result?.subscribers.length).to.be.eql(subscribers.length);
+  expect(result?.subscribers).to.have.members(subscribers.map((subscriber) => subscriber.subscriberId));
```
Comment (p-fernandez) [inline apps/api/src/app/events/usecases/map-trigger-recipients/map-trigger-recipients.e2e.ts:618]: Linter.

File: apps/api/src/app/events/usecases/map-trigger-recipients/map-trigger-recipients.e2e.ts [hunk 5]
```diff
@@ -171,10 +178,12 @@ describe('MapTriggerRecipientsUseCase', () => {
 
   describe('When feature enabled', () => {
     beforeEach(() => {
+      process.env.LAUNCH_DARKLY_SDK_KEY = '';
       process.env.FF_IS_TOPIC_NOTIFICATION_ENABLED = 'true';
     });
 
     afterEach(() => {
+      process.env.LAUNCH_DARKLY_SDK_KEY = originalLaunchDarklySdkKey;
```
Comment (p-fernandez) [inline apps/api/src/app/events/usecases/map-trigger-recipients/map-trigger-recipients.e2e.ts:186]: We restore the Launch Darkly SDK key when done with the tests to not influence other test suites.

File: apps/api/src/app/events/e2e/trigger-event-topic.e2e.ts [hunk 6]
```diff
@@ -288,8 +292,8 @@ describe('Topic Trigger Event', () => {
           channel: ChannelTypeEnum.SMS,
         });
 
-        expect(message._subscriberId.toString()).to.be.eql(subscriber._id);
-        expect(message.phone).to.equal(subscriber.phone);
+        expect(message?._subscriberId.toString()).to.be.eql(subscriber._id);
+        expect(message?.phone).to.equal(subscriber.phone);
```
Comment (p-fernandez) [inline apps/api/src/app/events/e2e/trigger-event-topic.e2e.ts:296]: Linter.

File: apps/api/src/app/feature-flags/feature-flags.module.ts [hunk 7]
```diff
@@ -8,13 +8,13 @@ import { USE_CASES } from './use-cases';
...
-const providers = USE_CASES;
 const services = [LaunchDarklyService, FeatureFlagsService];
+const providers = [...USE_CASES, ...services];
 
 @Module({
   imports: [SharedModule, AuthModule],
-  providers: [...providers, ...services],
-  exports: services,
+  providers,
+  exports: providers,
```
Comment (p-fernandez) [inline apps/api/src/app/feature-flags/feature-flags.module.ts:17]: I needed to export the use cases for being used in the Events module.

File: apps/api/src/app/feature-flags/e2e/get-feature-flag.e2e.ts [hunk 8]
```diff
@@ -72,13 +88,26 @@ describe('Get Feature Flag', () => {
...
+      });
+
+      describe('IS_TOPIC_NOTIFICATION_ENABLED', () => {
+        it(`should get the feature flag value stored in Launch Darkly (enabled)
+           when the SDK key env variable is set regardless of the feature flag set`, async () => {
+          process.env.FF_IS_TOPIC_NOTIFICATION_ENABLED = 'false';
 
-        const result = await getFeatureFlag.isTemplateStoreEnabled(featureFlagCommand);
+          const result = await getFeatureFlag.isTopicNotificationEnabled(featureFlagCommand);
```
Comment (p-fernandez) [inline apps/api/src/app/feature-flags/e2e/get-feature-flag.e2e.ts:107]: He... 😅

**Review by scopsy — APPROVED** [review]

**Review by LetItRock — APPROVED** [review]
🎖️

**Pull request merged**